controllers/etcd_controller_test.go

```diff
@@ -602,6 +605,214 @@ var _ = Describe("Druid", func() {
 	)
 })
 
+var _ = Describe("Multinode ETCD", func() {
+	//Reconciliation of new etcd resource deployment without any existing statefulsets.
+	Context("when adding etcd resources", func() {
+		var (
+			err      error
+			instance *druidv1alpha1.Etcd
+			sts      *appsv1.StatefulSet
+			svc      *corev1.Service
+			c        client.Client
+		)
+
+		BeforeEach(func() {
+			instance = getEtcd("foo07", "default", false)
+			c = mgr.GetClient()
+			ns := corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: instance.Namespace,
+				},
+			}
+			_, err = controllerutil.CreateOrUpdate(context.TODO(), c, &ns, func() error { return nil })
+			Expect(err).To(Not(HaveOccurred()))
+
+			storeSecret := instance.Spec.Backup.Store.SecretRef.Name
+			errors := createSecrets(c, instance.Namespace, storeSecret)
+			Expect(len(errors)).Should(BeZero())
+		})
+		It("no statefulsets are created when ETCD replicas are even number", func() {
+			// Update replicas in ETCD resource with 0
+			instance.Spec.Replicas = 0
+			Expect(c.Create(context.TODO(), instance)).To(Succeed())
+
+			Eventually(func() error {
+				return c.Get(context.TODO(), types.NamespacedName{
+					Name:      instance.Name,
+					Namespace: instance.Namespace,
+				}, instance)
+			}, timeout, pollingInterval).Should(BeNil())
+
+			sts = &appsv1.StatefulSet{}
+			// No StatefulSet has been created by controller as even number of replicas are not allowed
+			Eventually(func() error {
+				return c.Get(context.TODO(), types.NamespacedName{
+					Name:      instance.Name,
+					Namespace: instance.Namespace,
+				}, sts)
+			}, timeout, pollingInterval).Should(matchers.BeNotFoundError())
+
+			// Update replicas in ETCD resource with even number
+			Expect(kutil.TryUpdate(context.TODO(), retry.DefaultBackoff, c, instance, func() error {
+				instance.Spec.Replicas = 4
+				return nil
+			})).To(Succeed())
+
+			Eventually(func() error {
+				return c.Get(context.TODO(), types.NamespacedName{
+					Name:      instance.Name,
+					Namespace: instance.Namespace,
+				}, instance)
+			}, timeout, pollingInterval).Should(BeNil())
+
+			// No StatefulSet has been created by controller as even number of replicas are not allowed
+			sts = &appsv1.StatefulSet{}
+			Eventually(func() error {
+				return c.Get(context.TODO(), types.NamespacedName{
+					Name:      instance.Name,
+					Namespace: instance.Namespace,
+				}, sts)
+			}, timeout, pollingInterval).Should(matchers.BeNotFoundError())
+
+			// No Service has been created by controller as even number of replicas are not allowed
+			svc = &corev1.Service{}
+			Expect(c.Get(context.TODO(), types.NamespacedName{
+				Name:      fmt.Sprintf("%s-client", instance.Name),
+				Namespace: instance.Namespace,
+			}, svc)).Should(matchers.BeNotFoundError())
+			svc = nil
+		})
+		It("configmaps are mounted properly when ETCD replicas are odd number", func() {
```

Comment:
Would it be possible to re-use the common code between the case for `replicas: 1` and `replicas: 3`?

Reply:
Yes. Used describetable